File: src/chain-observers/collectors/common/block-header-collector.ts
import { BlockHeaderClassifier } from "../../classifiers";
import { TRegisterEvaluationContext } from "../../classifiers/collector-classifier-base";
import { DynamicGlobalPropertiesClassifier } from "../../classifiers/dynamic-global-properties-classifier";
import { TCollectorEvaluationContext } from "../../factories/data-evaluation-context";
import { CollectorBase, TAvailableClassifiers } from "../collector-base";

export class BlockHeaderCollector extends CollectorBase<BlockHeaderClassifier> {
  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [DynamicGlobalPropertiesClassifier];
  }

  public async get(data: TCollectorEvaluationContext) {
    const { headBlockNumber, currentWitness, headBlockTime, headBlockId } = await data.get(DynamicGlobalPropertiesClassifier);

    return {
      /*
       * Instruct TypeScript typings that BlockHeaderClassifier.name is actualy a Classifier name we expect.
       * This is required for the bundlers to properly deduce the type of the classifier in data evaluation context.
       */
      [BlockHeaderClassifier.name as "BlockHeaderClassifier"]: {
        number: headBlockNumber,
        timestamp: headBlockTime,
        witness: currentWitness,
        id: headBlockId
      } as TAvailableClassifiers["BlockHeaderClassifier"]
    };
  };
}

File: python/workerbee/chain_observers/collectors/common/block_header_collector.py
"""BlockHeaderCollector — derives block header from DynamicGlobalPropertiesClassifier.

Mirrors src/chain-observers/collectors/common/block-header-collector.ts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ...classifiers.block_header_classifier import BlockHeaderClassifier
from ...classifiers.collector_classifier_base import TRegisterEvaluationContext
from ...classifiers.dynamic_global_properties_classifier import DynamicGlobalPropertiesClassifier
from ..collector_base import CollectorBase

if TYPE_CHECKING:
    from ...factories.data_evaluation_context import TCollectorEvaluationContext


class BlockHeaderCollector(CollectorBase):
    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        return [DynamicGlobalPropertiesClassifier]

    async def get(self, data: TCollectorEvaluationContext) -> dict[str, Any]:
        dgp = await data.get(DynamicGlobalPropertiesClassifier)

        return {
            BlockHeaderClassifier.__name__: {
                "number": dgp["head_block_number"],
                "timestamp": dgp["head_block_time"],
                "witness": dgp["current_witness"],
                "id": dgp["head_block_id"],
            },
        }
